Sources/GRPC/AsyncAwaitSupport/GRPCAsyncBidirectionalStreamingCall.swift

```diff
@@ -70,93 +73,43 @@ public struct GRPCAsyncBidirectionalStreamingCall<Request, Response> {
-    self.responses = .init(responseStream)
+    self.responseSource = PassthroughMessageSource<Response, Error>()
+    self.responses = .init(PassthroughMessageSequence(consuming: self.responseSource))
+    self.requestStream = call.makeRequestStreamWriter()
   }
 
   /// We expose this as the only non-private initializer so that the caller
   /// knows that invocation is part of initialisation.
   internal static func makeAndInvoke(call: Call<Request, Response>) -> Self {
-    Self(call: call)
-  }
-
-  // MARK: - Requests
-
-  /// Sends a message to the service.
-  ///
-  /// - Important: Callers must terminate the stream of messages by calling `sendEnd()`.
-  ///
-  /// - Parameters:
-  ///   - message: The message to send.
-  ///   - compression: Whether compression should be used for this message. Ignored if compression
-  ///     was not enabled for the RPC.
-  public func sendMessage(
-    _ message: Request,
-    compression: Compression = .deferToCallDefault
-  ) async throws {
-    let compress = self.call.compress(compression)
-    let promise = self.call.eventLoop.makePromise(of: Void.self)
-    self.call.send(.message(message, .init(compress: compress, flush: true)), promise: promise)
-    // TODO: This waits for the message to be written to the socket. We should probably just wait for it to be written to the channel?
-    try await promise.futureResult.get()
-  }
-
-  /// Sends a sequence of messages to the service.
-  ///
-  /// - Important: Callers must terminate the stream of messages by calling `sendEnd()`.
-  ///
-  /// - Parameters:
-  ///   - messages: The sequence of messages to send.
-  ///   - compression: Whether compression should be used for this message. Ignored if compression
-  ///     was not enabled for the RPC.
-  public func sendMessages<S>(
-    _ messages: S,
-    compression: Compression = .deferToCallDefault
-  ) async throws where S: Sequence, S.Element == Request {
-    let promise = self.call.eventLoop.makePromise(of: Void.self)
-    self.call.sendMessages(messages, compression: compression, promise: promise)
-    try await promise.futureResult.get()
-  }
+    let asyncCall = Self(call: call)
+
+    asyncCall.call.invokeStreamingRequests { error in
+      asyncCall.responseParts.handleError(error)
+      asyncCall.responseSource.finish(throwing: error)
+    } onResponsePart: { responsePart in
+      // Handle the metadata, trailers and status.
+      asyncCall.responseParts.handle(responsePart)
+
+      // Handle the response messages and status.
+      switch responsePart {
+      case .metadata:
+        ()
+
+      case let .message(response):
+        // TODO: when we support backpressure we will need to stop ignoring the return value.
+        _ = asyncCall.responseSource.yield(response)
+
+      case let .end(status, _):
+        if status.isOk {
+          asyncCall.responseSource.finish()
+        } else {
+          asyncCall.responseSource.finish(throwing: status)
+        }
+      }
+    }
```

Comment:
Is there a way for us to share this implementation with the `GRPCAsyncServerStreamingCall` further down?